paasta_tools/autoscaling/autoscaling_service_lib.py

```diff
@@ -168,17 +181,56 @@ def http_metrics_provider(marathon_service_config, marathon_tasks, mesos_tasks,
     utilization = []
     for task in marathon_tasks:
         try:
-            endpoint_utilization = requests.get(
-                'http://%s:%s/%s' % (task.host, task.ports[0], endpoint),
-                headers={'User-Agent': get_user_agent()}
-            ).json()['utilization']
-            utilization.append(float(endpoint_utilization))
+            utilization.append(json_mapper(get_json_body_from_service(task.host, task.ports[0], endpoint)))
         except Exception:
             pass
     if not utilization:
         raise MetricsProviderNoDataError('Couldn\'t get any data from http endpoint %s for %s.%s' % (
             endpoint, marathon_service_config.service, marathon_service_config.instance))
-    return sum(utilization) / len(utilization)
+    return average(utilization)
+
+
+@register_autoscaling_component('uwsgi', SERVICE_METRICS_PROVIDER_KEY)
+def uwsgi_metrics_provider(marathon_service_config, marathon_tasks, endpoint='uwsgi_stats', *args, **kwargs):
+    """
+    Gets the average utilization of a service across all of its tasks, where
+    the utilization of a task is the percentage of non-idle workers as read
+    from the UWSGI stats endpoint
+
+    :param marathon_service_config: the MarathonServiceConfig to get data from
+    :param marathon_tasks: Marathon tasks to get data from
+    :param endpoint: The http endpoint to get the uwsgi stats from
+
+    :returns: the service's average utilization, from 0 to 1
+    """
+
+    def uwsgi_mapper(json):
+        workers = json['workers']
+        utilization = [1.0 if worker['status'] != 'idle' else 0.0 for worker in workers]
+        return average(utilization)
```

Comment:
There are two subtle points here that you are missing:

The first is that during uwsgi startup (and i believe in the case where the _master_ reloads), all workers may not have spawned yet. In order to avoid a race condition, you may want to check that all workers are in the `accepting` state before you can say anything conclusively about the utilization

```python
for worker in workers:
    if worker['accepting'] == 0:
        ...do something to indicate that it isn't ready
```

The second subtlety is that uwsgi has the concept of async "[cores](http://uwsgi-docs.readthedocs.io/en/latest/Async.html)", which will not contribute to your utilization calculation... To be honest, I don't know if they should. I really don't fully understand them (I think you can treat them as threads), but I think it is safe to ignore them for now.